File: 000401-000600/000435.无重叠区间/435.js
/**
 * @param {number[][]} intervals
 * @return {number}
 */

 // 方法一: 用时 232ms
 // 执行用时超过了 73%
 // 消耗内存超过了 47%
 // 难度: 中等
var eraseOverlapIntervals = function(intervals) {
  if (!intervals.length) {
    return 0
  }
  intervals = intervals.sort((a, b) => (a[1] - b[1]))
  let ret = 1
  let curLast = intervals[0][1]
  for (let i = 1; i < intervals.length; i++) {
    if (intervals[i][0] >= curLast) {
      ret ++
      curLast = intervals[i][1];
    }
  }
  return intervals.length - ret
};

console.log(eraseOverlapIntervals([ [3,4], [2,3], [1,2], [1,3] ]))
console.log(eraseOverlapIntervals([ [1,2], [1,3], [1,2] ]))
console.log(eraseOverlapIntervals([ [1,2], [2,3] ]))

File: 000401-000600/000435.无重叠区间/435.go
package main

import (
	"fmt"
	"sort"
)

// 方法一: 用时 224ms
// 执行用时超过了 51%
// 消耗内存超过了 54%
// 难度: 中等
func eraseOverlapIntervals(intervals [][]int) int {
	if len(intervals) == 0 {
		return 0
	}
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][1] < intervals[j][1]
	})
	ret := 1
	curLast := intervals[0][1]
	for i := 0; i < len(intervals); i++ {
		if intervals[i][0] >= curLast {
			ret ++
			curLast = intervals[i][1]
		}
	}
	return len(intervals) - ret
}

func main ()  {
	params := [][]int{[]int{3,4}, []int{2,3}, []int{1,2}, []int{1,3}}
	fmt.Println(eraseOverlapIntervals(params))
}
